Display an error text when ValueInput has invalid user input

jarolrod opened the pull request:
based on https://github.com/bitcoin-core/gui-qml/pull/255 and https://github.com/bitcoin-core/gui-qml/pull/256

This allows for a ValueInput to signal when it currently has an invalid input from the user and display an error text to the user. IntValidator values are essentially arbitrary right now until we get some more intelligent detection of system resources.

| invalid input | valid input |
| ------------- | ----------- |
| <img width="463" alt="Screen Shot 2023-02-10 at 7 08 22 PM" src="https://user-images.githubusercontent.com/23396902/218233776-059f03e3-a0ad-4c39-8548-32f8234f6e2d.png"> | <img width="463" alt="Screen Shot 2023-02-10 at 5 24 42 PM" src="https://user-images.githubusercontent.com/23396902/218233801-72206499-acc4-4bb6-bb04-18aa4ab7ba53.png"> |

Draft until we resolve the following concerns @GBKS @johnny9 @hebasto :
- The TextInput field can be left in an empty state if the user deletes all of the text and then clicks on another TextInput field. This doesn't write any value, but the TextInput should not be left empty
  - We could fill in the current value set in settings.json for the setting if the user leaves the TextInput field in an empty set, the stored value is what the GUI is going to use.
- Should the error text be shown in a new row, so that the TextInput field doesn't get moved around when showing/hiding the error text?


[![Windows](https://img.shields.io/badge/OS-Windows-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/win64/insecure_win_gui.zip?branch=pull/<PR>)
[![Intel macOS](https://img.shields.io/badge/OS-Intel%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/insecure_mac_gui.zip?branch=pull/<PR>)
[![Apple Silicon macOS](https://img.shields.io/badge/OS-Apple%20Silicon%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/insecure_mac_arm64_gui.zip?branch=pull/<PR>)
[![ARM64 Android](https://img.shields.io/badge/OS-Android-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/insecure_android_apk.zip?branch=pull/<PR>)

GBKS commented:
We talked about this a bit on Slack, so just repeating my thoughts (IIRC). If we evaluate the validity of the input on each key stroke, it's pretty annoying if the layout changes. So it's better to evaluate it when the input field loses focus, and in that case, the layout reflow should be OK.

When entering nothing, a reset to the current value also makes sense.

A more helpful error text would be "Enter a value between X and Y".

Is it possible to use MB instead of MiB? I'd assume that more people know what MB are and might confuse MiB with the Men in Black.

![](https://media.giphy.com/media/NwgX4KhTTTpDi/giphy.gif)

hebasto commented:
> based on #255

It has just been merged. Rebase?

jarolrod commented:
Updated from[ 1e89928 ](https://github.com/bitcoin-core/gui-qml/commit/1e8992805459b275c5565306c8235b8cbcbf550d)to[ 00fac08](https://github.com/bitcoin-core/gui-qml/commit/00fac088265af0e0c94499e619de44f32a24d1fe), [compare](https://github.com/bitcoin-core/gui-qml/compare/1e8992805459b275c5565306c8235b8cbcbf550d..00fac088265af0e0c94499e619de44f32a24d1fe)

Changes: rebased to resolve conlicts

jarolrod commented:
Updated from[ 00fac08 ](https://github.com/bitcoin-core/gui-qml/commit/00fac088265af0e0c94499e619de44f32a24d1fe)to [8bfafde](https://github.com/bitcoin-core/gui-qml/commit/8bfafdecdcf51430b3745daad1c1bb8adab465bb), [compare](https://github.com/bitcoin-core/gui-qml/compare/00fac088265af0e0c94499e619de44f32a24d1fe..8bfafdecdcf51430b3745daad1c1bb8adab465bb)

Changes: Addressed Review [feedback](https://github.com/bitcoin-core/gui-qml/pull/257#issuecomment-1427506593)
- Validation happens on editing finished instead of every input keystroke
- Display informative error text with what the valid values are


TODO in follow-ups:
- Restrict the characters that can be input into the field, best option is [inputMask](https://doc.qt.io/qt-6/qml-qtquick-textinput.html#inputMask-prop)
- Restrict the width of the error text
- Address review feedback that the default value or the value that options model has for a setting should be set if the user leaves a textinput in an empty state

hebasto commented:
> based on https://github.com/bitcoin-core/gui-qml/pull/271

https://github.com/bitcoin-core/gui-qml/pull/271 has just been merged.

jarolrod commented:
Updated from [8bfafde ](https://github.com/bitcoin-core/gui-qml/commit/8bfafdecdcf51430b3745daad1c1bb8adab465bb)to [d8cf8d0](https://github.com/bitcoin-core/gui-qml/commit/d8cf8d018a9dcb7efa541df3e4b421649b2644b5), [compare](https://github.com/bitcoin-core/gui-qml/compare/8bfafdecdcf51430b3745daad1c1bb8adab465bb..d8cf8d018a9dcb7efa541df3e4b421649b2644b5)

Changes: rebased over main

GBKS commented:
Just tested it and I think the input validation needs a bit more work. While it worked well when I entered a low-ish number, I was able to instantly break the app by entering random text (or a really high number like 100 trillion).

<img width="752" alt="image" src="https://user-images.githubusercontent.com/695901/221504348-8b36a6a5-e5e4-464d-9b09-13a67b731804.png">

hebasto commented:
> based on https://github.com/bitcoin-core/gui-qml/pull/256

It has been just merged.

jarolrod commented:
Updated from from [d8cf8d0 ](https://github.com/bitcoin-core/gui-qml/commit/d8cf8d018a9dcb7efa541df3e4b421649b2644b5)to [a51820f](https://github.com/bitcoin-core/gui-qml/commit/a51820f9e704b36aff0e67695b33201dd85310ea), [compare](https://github.com/bitcoin-core/gui-qml/compare/d8cf8d018a9dcb7efa541df3e4b421649b2644b5..a51820f9e704b36aff0e67695b33201dd85310ea)

changes: rebased over main

jarolrod commented:
Updated from [a51820f ](https://github.com/bitcoin-core/gui-qml/commit/a51820f9e704b36aff0e67695b33201dd85310ea)to [524decd](https://github.com/bitcoin-core/gui-qml/commit/524decd192d45192933c26b8e665a7565d78e6ab), [compare](https://github.com/bitcoin-core/gui-qml/compare/a51820f9e704b36aff0e67695b33201dd85310ea..524decd192d45192933c26b8e665a7565d78e6ab)

Changes: @GBKS your concerns should now be addressed here. I hadn't addressed those two points yet because I was thinking about what changes would be needed to ValueInput to allow for it to be used within the Proxy screen, but I'll leave that thinking to another PR.

- Set a maximum input length of 5 to the ValueInput control. This is enough to enter a valid value for dbcache, threads, and prune target
- Set an IntValidator without specifying the valid bottom or top values. What this does is it prevents any characters that are not numeric from being entered, while allowing a character like `-` to enter negative numbers (can be used in threads).

hebasto commented:
Merged as a part of #245.
